gnome-image-installer/util/eos-reformatter.c

```diff
@@ -0,0 +1,972 @@
+                         NULL  /* default value */,
+                         G_PARAM_CONSTRUCT_ONLY | G_PARAM_READWRITE);
+
+  _props[PROP_DEVICE] =
+    g_param_spec_string ("device",
+                         "Device",
+                         "Disk device to write the image to",
+                         NULL  /* default value */,
+                         G_PARAM_CONSTRUCT_ONLY | G_PARAM_READWRITE);
+
+  _props[PROP_PROGRESS] =
+    g_param_spec_double ("progress",
+                         "Progress",
+                         "Progress of the reformatting",
+                         0.0  /* minimum value */,
+                         1.0  /* maximum value */,
+                         0.0  /* default value */,
+                         G_PARAM_READWRITE);
+
+  _signals[SIG_FINISHED] =
+    g_signal_newv ("finished",
+                   G_TYPE_FROM_CLASS (gobject_class),
+                   G_SIGNAL_RUN_LAST | G_SIGNAL_NO_RECURSE | G_SIGNAL_NO_HOOKS,
+                   NULL /* closure */,
+                   NULL /* accumulator */,
+                   NULL /* accumulator data */,
+                   NULL /* C marshaller */,
+                   G_TYPE_NONE /* return_type */,
+                   0     /* n_params */,
+                   NULL  /* param_types */);
+
+  g_object_class_install_properties (gobject_class,
+                                     N_PROPERTIES,
+                                     _props);
+}
+
+EosReformatter *
+eos_reformatter_new (const gchar *image, const gchar *signature, const gchar *device)
+{
+  return EOS_REFORMATTER (g_object_new (EOS_TYPE_REFORMATTER,
+                                      "image", image,
+                                      "device", device,
+                                      "signature", signature,
+                                      NULL));
+}
+
+/* Private API */
+
+static void
+eos_reformatter_maybe_finish (EosReformatter *reformatter)
+{
+  if (reformatter->finished)
+    return;
+
+  if (reformatter->error == NULL && reformatter->gpg > 0)
+    return;
+
+  if (reformatter->error == NULL && reformatter->progress < 1.0)
+    return;
+
+  reformatter->finished = TRUE;
+
+  g_signal_emit (reformatter, _signals[SIG_FINISHED], 0);
+}
+
+static EosAlignedBuffer *
+eos_reformatter_get_free_buffer (EosReformatter *reformatter, gboolean exhaust)
+{
+  /* If the exhaust flag is not set, we only allow half of the buffers to be
+   * reserved. This is mainly to avoid reading queue to hog free buffers.
+   */
+  int half_buffers = reformatter->total_buffers / 2;
+  EosAlignedBuffer *buf = NULL;
+
+  if (exhaust || g_async_queue_length (reformatter->free_queue) < half_buffers)
+    {
+      buf = g_async_queue_try_pop (reformatter->free_queue);
+    }
+
+  /* A free buffer was available, return it */
+  if (buf != NULL)
+    {
+      buf->len = 0;
+      return buf;
+    }
+
+  /* Initialize our pool, if it's not created yet */
+  if (reformatter->pool == NULL)
```

Comment:
Since the pool might be large, it would be awkward if the reformatter is created ahead of time and it would immediately gulp a large chunk of memory. Not sure how relevant the use-case is though.